### PR TITLE
create empty fakes

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- substitute fake features with empty arrays and use attribute mask to rebuild original shapes
- part of https://github.com/RasaHQ/rasa/issues/6670
**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/policies/ted_policy.py [hunk 1]
```diff
@@ -819,13 +819,20 @@ def _encode_features_per_attribute(
         Returns:
             A tensor combining  all features for `attribute`
         """
-        attribute_mask = tf_batch_data[attribute][MASK][0]
+        # attribute_mask = tf_batch_data[attribute][MASK][0]
```
Comment (tabergma) [inline rasa/core/policies/ted_policy.py]: if not needed, it can be removed

File: rasa/core/policies/ted_policy.py [hunk 2]
```diff
@@ -819,13 +819,20 @@ def _encode_features_per_attribute(
...
 
         if attribute in SEQUENCE_FEATURES_TO_ENCODE:
             _sequence_lengths = tf.cast(
```
Comment (tabergma) [inline rasa/core/policies/ted_policy.py:822]: should we add a comment that sequence length contains 0 length if no features for this attribute are present?

File: rasa/core/policies/ted_policy.py [hunk 3]
```diff
@@ -919,18 +929,37 @@ def _convert_to_original_shape(
         # mapping the values of attribute features to the position in the resulting
         # tensor.
 
-        batch_dim = tf.size(dialogue_lengths)
-        dialogue_dim = tf.reduce_max(dialogue_lengths)
+        batch_dim = tf.shape(attribute_mask)[0]
+        dialogue_dim = tf.shape(attribute_mask)[1]
         units = attribute_features.shape[-1]
+        if dialogue_lengths is None:
```
Comment (tabergma) [inline rasa/core/policies/ted_policy.py:932]: should we add a comment that we assume that dialogue length is 1 in case it is not provided?

File: rasa/core/policies/ted_policy.py [hunk 4]
```diff
@@ -909,7 +920,6 @@ def _convert_to_original_shape(
         """
         # dialogue lengths contains the actual dialogue length
         # shape is batch-size x 1
```
Comment (tabergma) [inline rasa/core/policies/ted_policy.py:919]: should we move this to the doc string?

File: rasa/utils/tensorflow/model_data.py [hunk 5]
```diff
@@ -1039,6 +1039,18 @@ def _create_label_ids(label_ids: FeatureArray) -> np.ndarray:
 
         raise ValueError("Unsupported label_ids dimensions")
 
+    @staticmethod
+    def _filter_4d_arrays(array_of_array_of_features):
```
Comment (tabergma) [inline rasa/utils/tensorflow/model_data.py:1043]: please add types

File: rasa/utils/tensorflow/model_data.py [hunk 6]
```diff
@@ -1039,6 +1039,18 @@ def _create_label_ids(label_ids: FeatureArray) -> np.ndarray:
...
+        return list(
+            filter(
+                lambda x: len(x) > 0,
+                [
+                    list(filter(lambda x: x.shape[0] > 0, array_of_features))
```
Comment (tabergma) [inline rasa/utils/tensorflow/model_data.py:1048]: can't we use the parameter `number_of_dimension`?

**Review by tabergma — COMMENTED** [review]
Will take a closer look at the resulting shapes now.

File: rasa/utils/tensorflow/model_data.py (hunk 6 above)
Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:1050]: what do you mean? `number_of_dimension` is always `4` here, it's just sequence dim might be `0`

Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:1050]: ah, misread the function name, can you please add a comment? thanks

**Comment by Ghostvv** [comment]
@tabergma I fixed the relevant tests. Could you please review

File: rasa/utils/tensorflow/model_data.py [hunk 7]
```diff
@@ -1039,6 +1039,23 @@ def _create_label_ids(label_ids: FeatureArray) -> np.ndarray:
 
         raise ValueError("Unsupported label_ids dimensions")
 
+    @staticmethod
+    def _filter_4d_arrays(
```
Comment (tabergma) [inline rasa/utils/tensorflow/model_data.py:1043]: should we rename this to `filter_out_fake_inputs`?

**Review by tabergma — COMMENTED** [review]
I think we should add a comment to explain the difference between fake and real features. Will check out the branch in a sec and take a closer look at the code itself.

**Comment by tabergma** [comment]
There seems to be an issue with entity training in DIET, not sure if that is related to your changes, but something we need to double check.
```
Epochs: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:26<00:00,  7.50it/s, t_loss=1.156, i_acc=0.986, e_f1=0.545]
```

**Comment by Ghostvv** [comment]
on `e2e` there is a issue with starting training DIET, my PR fixed it (non intentionally), but maybe it is related?

File: rasa/utils/tensorflow/model_data_utils.py [hunk 8]
```diff
@@ -434,8 +427,9 @@ def _extract_features(
             dense_features[key].append(value)
 
         # add additional dimensions to attribute mask to get a 3D vector
-        # resulting shape dialogue length x 1 x 1
-        attribute_mask = np.expand_dims(np.expand_dims(attribute_mask, -1), -1)
+        # resulting shape dialogue length x 1
+        # but keep attribute_mask in 3d: (batch x dialogue_len x 1)
```
Comment (tabergma) [inline rasa/utils/tensorflow/model_data_utils.py:431]: this comment is a bit confusing, `resulting shape` of what?

**Comment by tabergma** [comment]
Training DIET on `e2e` branch results in the same. So most likely something that was introduced before. I'll add it to the list of things we need to do. Nothing that should be fixed within this PR I guess.

**Review by tabergma — APPROVED** [review]
Tensors and their shapes look good 🚀

File: rasa/utils/tensorflow/model_data_utils.py (hunk 8 above)
Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data_utils.py]: rephrased

File: rasa/core/policies/ted_policy.py [hunk 9]
```diff
@@ -808,7 +815,38 @@ def _emebed_dialogue(
 
     def _encode_features_per_attribute(
         self, tf_batch_data: Dict[Text, Dict[Text, List[tf.Tensor]]], attribute: Text
-    ) -> Optional[tf.Tensor]:
+    ) -> tf.Tensor:
+
+        return tf.cond(
```
[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/core/policies/ted_policy.py:836]: created comments

**Comment by Ghostvv** [comment]
I created comments about what "fake" features are, could you please take a look?

File: rasa/core/policies/ted_policy.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/core/policies/ted_policy.py:836]: thanks, looks good
